File: holoken-core/src/main/java/com/holokenmod/creation/GridCreator.java
package com.holokenmod.creation;

import com.holokenmod.RandomSingleton;
import com.holokenmod.Randomizer;
import com.holokenmod.creation.cage.GridCageCreator;
import com.holokenmod.grid.Grid;
import com.holokenmod.options.DifficultySetting;
import com.holokenmod.options.GameVariant;

public class GridCreator {
	private final GameVariant variant;
	private final Randomizer randomizer;
	private final PossibleDigitsShuffler shuffler;
	
	public GridCreator(final GameVariant variant) {
		this(RandomSingleton.getInstance(), new RandomPossibleDigitsShuffler(), variant);
	}
	
	public GridCreator(final Randomizer randomizer, final PossibleDigitsShuffler shuffler, final GameVariant variant) {
		this.randomizer = randomizer;
		this.shuffler = shuffler;
		this.variant = variant;
	}
	
	public Grid createRandomizedGridWithCages() {
		randomizer.discard();
		
		Grid newGrid;
		
		do {
			newGrid = new Grid(variant);
			
			newGrid.addAllCells();
			
			randomiseGrid(newGrid);
			createCages(newGrid);
		} while (!isWantedDifficulty(newGrid));
		
		return newGrid;
	}
	
	private boolean isWantedDifficulty(Grid grid) {
		if (variant.getOptions().getDifficultySetting() == DifficultySetting.ANY) {
			return true;
		}
		
		GridDifficultyCalculator calculator = new GridDifficultyCalculator(grid);
		
		if (!calculator.isGridVariantSupported()) {
			return true;
		}
		
		return calculator.getDifficulty() == variant.getOptions().getDifficultySetting().getGameDifficulty();
	}
	
	private void createCages(Grid grid) {
		GridCageCreator creator = new GridCageCreator(randomizer, grid);
		
		creator.createCages();
	}
	
	private void randomiseGrid(Grid grid) {
		GridRandomizer randomizer = new GridRandomizer(shuffler, grid);
		
		randomizer.createGrid();
	}
}

File: holoken-core/src/main/kotlin/com/holokenmod/creation/GridCreator.kt
package com.holokenmod.creation

import com.holokenmod.RandomSingleton
import com.holokenmod.Randomizer
import com.holokenmod.creation.cage.GridCageCreator
import com.holokenmod.grid.Grid
import com.holokenmod.options.DifficultySetting
import com.holokenmod.options.GameVariant

class GridCreator(
    private val variant: GameVariant,
    private val randomizer: Randomizer = RandomSingleton.instance,
    private val shuffler: PossibleDigitsShuffler = RandomPossibleDigitsShuffler()
) {
    fun createRandomizedGridWithCages(): Grid {
        randomizer.discard()
        var newGrid: Grid
        do {
            newGrid = Grid(variant)
            newGrid.addAllCells()
            randomiseGrid(newGrid)
            createCages(newGrid)
        } while (!isWantedDifficulty(newGrid))

        return newGrid
    }

    private fun isWantedDifficulty(grid: Grid): Boolean {
        if (variant.options.difficultySetting == DifficultySetting.ANY) {
            return true
        }
        val calculator = GridDifficultyCalculator(grid)
        return if (!calculator.isGridVariantSupported) {
            true
        } else {
            calculator.difficulty == variant.options.difficultySetting.gameDifficulty
        }
    }

    private fun createCages(grid: Grid) {
        val creator = GridCageCreator(randomizer, grid)
        creator.createCages()
    }

    private fun randomiseGrid(grid: Grid) {
        val randomizer = GridRandomizer(shuffler, grid)
        randomizer.createGrid()
    }
}
